### ISSUE TITLE
Add tools for HTTP Fuzzing

### ISSUE THREAD

**Issue by pimterry** [httptoolkit/feedback#49]
In addition to #48, a useful extension would be to send _fuzzed_ requests. 

I.e. not just send this one request, but define a request generator that sends a bunch of different requests, trying different edge cases, to see what happens and find bugs/security issues.

---

_Does this affect you too? Click the +:smiley: icon in the top right and add a :+1: on issues to help decide where HTTP Toolkit goes next, or [go take a look & vote on the most popular ideas so far](https://github.com/httptoolkit/feedback/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc)._

**Comment by dmknght** [comment]
Hi! I am having a few things about fuzzing idea:
- It must be fast, lightweight, good performance and easy to customize requests
- It must be very easy to check result
- It should be modular. I am thinking we can use something like yara rules to analysis responses and match kind of common vulnerabilities (a scanner like and as far as i remember, there is no web fuzzing tool can do it or can do it good)

**Comment by pimterry** [comment]
Thanks @dmknght! Do you have any examples of other similar tools like this? Can you be more specific on how you would want to customize the requests?

For checking the result, my plan is:

* Record all requests (as HTTP Toolkit does with any other request)
* Add complex filtering tools  (see #27), so you can filter for (for example) status !== 200 & body contains "SQL error"

It's probably not practical to add support for yara rules themselves, but it would definitely be possible to have some kind of import/export support for filtering rules, so you can save & reuse these easily. Would that work for you?

**Comment by dmknght** [comment]
> Do you have any examples of other similar tools like this

I don't. Actually i wrote something in python for that kind of tool but i have only few lines of code and i have no time to continue it. (CLI only so a gui like your tool would be better)

> Do you have any examples of other similar tools like this

Like custom cookies and header parameters. The fuzzing could be replace any position in headers as same as body i think. 

> so you can filter for (for example) status !== 200 & body contains "SQL error"

I think a good filter plugin can do this easily: responses have headers and http response, so users just need to do something like `response.status_code !== 200 and "SQL error" in response.text`. Maybe using yara is not the best way but you have to create something easy for create and read plugins. If it is possible so i think it is better than calling a sub process or something like that :D

**Comment by dmknght** [comment]
@pimterry I am having some more about it:
- controller: control threads number, delay, injection position. The injection position can be 1 or multiple place (header, parameter values) and select payload list.
- analyzer which should be modular, easy to use scripting. It takes requests, responses and payloads from controller and analysis with modules inside it. For example: Xss -> check if unique string (contains special characters like `<` `>`, ...) is not encoded and stored in response -> show xss; response has strings like "MySQL state..." -> SQL  Injection.
- UI: display useful information like: response code, title, page diff, ... When users click / double click on a line in table, the tool will display full requests and response.
What do you think about this model?

**Comment by pimterry** [comment]
I think the features you describe in 'UI' already exist, so that bit is easy :smile:.

The controller is similar to what I'm thinking. I'm expecting to build a normal postman-style HTTP client page soon (#48). This would be an extra feature there: in addition to sending just one request to `example.com/abc?message=test`, you can also send a set of requests (configuring delay & parallelism) to `example.com/${randomString()}?message=${xss()}`. Exact UX TBC, but some kind of nice UI to make inserting inline variables like this manageable. You could also add variables like these into header values too, or whatever else.

The analyzer part I'm more unsure about. Including a full scripting engine is much more complex, and I'm not clear why it's necessary. I'm intending to build search filters for requests anyway (#27). With those, you would easily be able to find responses where the body contains a given string, like "MySQL state" or "<script>", or do more complex checks. Imagine a UI sort-of like this: https://dribbble.com/shots/2718650-Rule-Based-Advanced-Search-with-Applied-Filters.

Do you have more complicated tests that you think this kind of filtering might not handle? I'd prefer to use the same filtering UI and logic when searching the requests to do this, if possible, to keep everything consistent from a UX point of view, and so I can reuse lots of the implementation.

**Comment by dmknght** [comment]
So i think controller and ui is pretty easy. The only problem is analyzer. I agree scripting is hard. I think search for string is kinda limited (even most of analysis process should use string base detection). My idea is create a useful tool can replace wfuzz and other web fuzzer. Here is my project that i mentioned before `https://github.com/dmknght/NGfuzz/`. You can see examples about rules (native python language, no 3rd language) in here `https://github.com/dmknght/NGfuzz/tree/master/modules/ActiveScan`. I think i can't find any other problem until we have features and test with test cases because it is kinda hard to image everything right now.
p/s: How about the memory management? Because fuzzing tasks can have hundred requests so i think data is kinda big.

**Comment by dmknght** [comment]
p/s: do we have a chance to analysis response time?

**Comment by pimterry** [comment]
> How about the memory management? Because fuzzing tasks can have hundred requests so i think data is kinda big.

I don't think this is a huge problem, although it depends on how big your responses are. Right now HTTP Toolkit can intercept thousands of requests in a session with no problem. The only real limitation is the total size of the request bodies.

What sort of responses are you receiving from normal fuzzing requests? If you're receiving lots of simple JSON or HTML responses that's probably not a problem. An average HTML file is about [25KB](https://httparchive.org/reports/page-weight#bytesHtml), so we can store 4,000 in memory with just 100MB available. We could improve this further by storing only the encoded response, rather than the decoded response, which would be significantly smaller (right now we decode when you first view the request body, and then cache that result until the request is cleared completely)

It could also be possible to automatically drop any responses that don't match your filters, so they disappear immediately. That would totally get rid of this problem, since we'd store just the matching responses, but I think it's probably ok for now, so we can leave that for later.

> Do we have a chance to analysis response time?

Response time is tracked, and available right now in the performance section. It would be easy to include that in the set of things you can filter on: e.g. to match requests with response duration > 100ms. Is that what you're thinking?

**Comment by dmknght** [comment]
> we can store 4,000 in memory with just 100MB available

That is impressed. I really don't like something eat memory like *java*

>  to match requests with response duration > 100ms. Is that what you're thinking?

Yes i think it is like that. For example: Blind SQL Injection has a payload that delay response (sleep) so it might be the key. Or from response at least we can see "something is wrong with this payload"

**Comment by pimterry** [comment]
> > we can store 4,000 in memory with just 100MB available

> That is impressed. I really don't like something eat memory like java

To be fair, I'm just talking about the memory usage of the requests themselves here, since that's the bit that changes depending on how much fuzzing you do. There's still the other overhead of the app itself, but that doesn't change, so it doesn't affect how much fuzzing we can handle.

> Blind SQL Injection has a payload that delay response (sleep) so it might be the key.

This makes lots of sense! Thanks, that's a really useful example.

I'm interested in doing this, but I'm not likely to have lots of time to start on it super soon. If you're keen, do you want to contribute? There's two places you could start:

* The [server](https://github.com/httptoolkit/httptoolkit-server) needs an endpoint to allow it to send requests (to power #48)
* The [UI](https://github.com/httptoolkit/httptoolkit-ui) needs filtering tools, to let you start filtering requests like we're discussing here (#27). 

Happy to talk more about the details of either if you're interested. I'm expecting to put the filtering in as a Pro feature, but I'd be happy to give you a year of Pro free if you make any real progress on either of those :smiley:.
